### PR TITLE
Properly map package source credentials to corresponding source

### PR DESCRIPTION
**Pull request by alpaix**
Fixes https://github.com/NuGet/Home/issues/2151.

The key idea behind this change was to eliminate use of CredentialStore
responsible for caching credentials per given host as opposed to the right
"per source" mapping. In order to achieve that package source credentials
needed to be associated with corresponding HttpSource instance.

For better clarity and granular testing two new message handlers were
factored out of HttpSource:
- HttpSourceAuthenticationHandler
- StsAuthenticationHandler

Decomissioned CredentialStore.

Introduced StsTokenStore in place of CredentialStore to maintain per host
mapping for STS tokens.

Removed PromptForCredential callback. Authentication handlers invoke
credential service directly.

Added authentication handlers tests

//cc @rrelyea @joelverhagen @emgarten @yishaigalatzer @zhili1208

### CODE REVIEW

File: src/NuGet.Clients/NuGet.CommandLine/Commands/Command.cs [hunk 1]
```diff
@@ -161,20 +161,12 @@ protected void SetDefaultCredentialProvider()
...
-            NuGet.Protocol.Core.v3.HttpHandlerResourceV3.PromptForCredentialsAsync =
-                async (uri, type, message, cancellationToken) => await credentialService.GetCredentialsAsync(
-                    uri,
-                    proxy: null,
-                    type: type,
-                    message: message,
-                    cancellationToken: cancellationToken);
-
-            NuGet.Protocol.Core.v3.HttpHandlerResourceV3.CredentialsSuccessfullyUsed = (uri, credentials) =>
+            NuGet.Protocol.HttpHandlerResourceV3.CredentialSerivce = credentialService;
```
Comment (joelverhagen) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/Command.cs]: nit: this may be a good time to fix the spelling mistake

Comment (alpaix) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/Command.cs]: I spent an hour solving a compile problem because of this typo!

**Comment by joelverhagen** [comment]
Good work on the unit tests. Is there any hope for integration tests for proxy and STS?

**Comment by alpaix** [comment]
@joelverhagen Need a dedicated setup for these. I tested the change with ad-hoc proxy and windows auth servers.

**Comment by joelverhagen** [comment]
We already have dedicated setup for 3rd party servers (the work @zhili1208 did), could these be added to the list?

**Comment by alpaix** [comment]
Both of the configurations (proxy and windows auth) require more than just a server:
- for proxy: local subnet, firewall, gateway, proxy server
- for windows auth: domain joined network with domain controller

**Comment by ygeelen-tb** [comment]
Do you know in which version of NuGet.exe and in which version of the Visual Studio NuGet extension this fix will be included?

**Comment by ygeelen-tb** [comment]
@alpaix is this fix included in the latest version available of the VS NuGet Package Manager extension?

**Comment by alpaix** [comment]
@ygeelen-tb The fix targets v3,6 planned for release in November this year.
If you want to evaluate it you can download the vsix from NuGet daily (unstable) feed [here](https://www.myget.org/F/nuget-vsix/vsix/NuGet.0d421874-a3b2-4f67-b53a-ecfce878063b-3.5.0.1580.vsix).

**Comment by ygeelen-tb** [comment]
Thanks, for when is v3.6 planned? Is there a roadmap timeline somewhere?

**Comment by ygeelen-tb** [comment]
@alpaix When is the v3.6 release planned? There's quite a big gap with the current version (2.8.60318.667) so it seems this could take a while...

**Comment by emgarten** [comment]
@ygleen-tb use 4.0.0 from http://dist.nuget.org/index.html

**Comment by ygeelen-tb** [comment]
Is nuget.exe 4.0.0 also used in VS 2015 VSIX 3.5 ?
Or should we upgrade to VS 2017 to have nuget.exe 4.0.0?

**Comment by ygeelen-tb** [comment]
@emgarten Will there be a VSIX 2015 with nuget.exe 4.0.0 released? If so, when?
Or do we need to upgrade to VS 2017 RC to get nuget.exe 4.0.0?

**Comment by emgarten** [comment]
You will need to upgrade to VS 2017

**Comment by ygeelen-tb** [comment]
@emgarten I did so now, but I can't even test it. See my comment in the following issue: https://developercommunity.visualstudio.com/content/problem/24733/visual-studio-nuget-package-manager-ignores-nugetc.html

It seems NuGet releases are part of VS and VS updates now, but I can't find nuget.exe 4.0.0 anywhere on my drive. Otherwise I'd try putting my config file next to it.

**Comment by ygeelen-tb** [comment]
@alpaix @yishaigalatzer We tested on VS 2017 now and we still get the dreaded Prompt window for entering credentials ??
It's supposed to read them from the file C:\Program Files (x86)\NuGet\NuGetDefaults.config

Has the XML schema of NuGetDefaults.config changed? We're using the packageSourceCredentials element to specify the credentials for each packageSource.
https://docs.microsoft.com/en-us/nuget/consume-packages/configuring-nuget-behavior#nuget-defaults-file

I also read about CredentialProviders (VSIX for Visual Studio or exe for nuget.exe) but I'd like to avoid implementing/maintaining one.. We're using ProGet.

**Comment by alpaix** [comment]
@ygeelen-tb Machine-wide config should be placed at `C:\Program Files (x86)\NuGet\Config\NuGetDefaults.config`. Please let me know if this works for you.

**Comment by ygeelen-tb** [comment]
Hi, I tested there and it's showing the feeds but prompts me for credentials instead of using the config credentials when accessing one.
Also once I enter credentials manually, it keeps sending them for all package sources (different URL, same domain though). There are different credentials for each package source in the config.

**Comment by ygeelen-tb** [comment]
I wrote a visual studio credential provider that reads in the config file but it's always reusing the credentials that were passed with the first accessed feed.. seems like visual studio is caching them?

**Comment by ygeelen-tb** [comment]
@alpaix @yishaigalatzer Any feedback on this?
I originally created the issue over 1 year ago https://github.com/NuGet/Home/issues/2151
It still doesn't seem like the packageSourceCredentials per feed are respected.

**Comment by ygeelen-tb** [comment]
I tested the CredentialProvider with Fiddler and it's indeed always sending the same credentials.
But instead I'd prefer to simply use NuGetDefaults.config if it would work.

**Comment by alpaix** [comment]
Could you please elaborate on how your sources and credentials are structured in `NuGetDefaults.config`?
What output are you seeing when running `nuget list -verbosity detailed`?

**Comment by ygeelen-tb** [comment]
Hi @alpaix,

Attached you can find a screenshot of our configuration. For company policy reasons I needed to erase some content. But the package source names do match.

To test the list command, I downloaded nuget.exe 4.0.0 into the solution. When executed, it prompted me for credentials as well like Visual Studio does.

nugetdefaults config
nugetlistcommandoutput

**Comment by alpaix** [comment]
@ygeelen-tb I was able to repro the problem locally and filed new issue on your behalf NuGet/Home#4938. Feel free to update the scenario if needed. We will triage it within next few days.

**Pull request merged**